src/domain/project/command.rs

```diff
@@ -170,10 +170,15 @@ async fn assert_permission(
     match credential {
         Credential::Auth0(user_id) => {
             let result = cache.find_user_permission(user_id, project_id).await?;
-            ensure!(result.is_some(), "user doesnt have permission");
+            if result.is_none() {
+                return Err(Error::PermissionDenied(
+                    "user doesnt have permission".into(),
+                ));
+            }
+
             Ok(())
         }
-        Credential::ApiKey(_) => bail!("rpc doesnt support api-key"),
+        Credential::ApiKey(_) => Err(Error::PermissionDenied("rpc doesnt support api-key".into())),
```

Comment:
Here `Unauthenticated` as well.

Reply:
The credential is already valid and this assert will only guarantee that only Auth0 can use the project functions.